activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/ServiceTaskDelegateExpressionActivityBehavior.java

```diff
@@ -83,16 +85,21 @@ public void execute(DelegateExecution execution) {
         }
 
         Object delegate = DelegateExpressionUtil.resolveDelegateExpression(expression, execution, fieldDeclarations);
-        if (delegate instanceof ActivityBehavior) {
+        if (delegate instanceof DelegateExecutionFunction function) {
+            DelegateExecutionOutcome outcome = function.apply(execution);
+            if (outcome == DelegateExecutionOutcome.LEAVE_EXECUTION) {
+                leave(execution);
+            }
+        } else if (delegate instanceof ActivityBehavior) {
 
-          if (delegate instanceof AbstractBpmnActivityBehavior) {
-            ((AbstractBpmnActivityBehavior) delegate).setMultiInstanceActivityBehavior(getMultiInstanceActivityBehavior());
+          if (delegate instanceof AbstractBpmnActivityBehavior behavior) {
+            behavior.setMultiInstanceActivityBehavior(getMultiInstanceActivityBehavior());
           }
 
           Context.getProcessEngineConfiguration().getDelegateInterceptor().handleInvocation(new ActivityBehaviorInvocation((ActivityBehavior) delegate, execution));
 
-        } else if (delegate instanceof JavaDelegate) {
-          Context.getProcessEngineConfiguration().getDelegateInterceptor().handleInvocation(new JavaDelegateInvocation((JavaDelegate) delegate, execution));
+        } else if (delegate instanceof JavaDelegate javaDelegate) {
+          Context.getProcessEngineConfiguration().getDelegateInterceptor().handleInvocation(new JavaDelegateInvocation(javaDelegate, execution));
```

Comment:
How about using a switch here on the instance type?

Reply:
Good suggestion. Done!